Sources/SwiftProductGenerator.swift

```diff
@@ -0,0 +1,109 @@
+//
+//  SwiftProductGenerator.swift
+//
+//
+//  Created by Øyvind Grimnes on 09/09/15.
+//
+//
+
+
+/**
+A cartesian product generator
+
+:discussion:    Generates the cartesian product of collections, or repetitions
+                of a collection. The rightmost element is advanced every
+                iteration which ensures that if the input is sorted, the
+                output will be sorted as well.
+*/
+public struct Product<T, C: Collection>: IteratorProtocol, Sequence
+    where C.Iterator.Element == T
+{
+
+    /// Private variable to manage where to get next element in each pool
+    private var indices : [C.Index]
+    /// Pool containing all the collections to be combined
+    private var pools   : [C]
+    /// Terminate the generation on completion
+    private var done    : Bool = false
+
+
+    /**
+    Initiates the generator with an array of collections to be combined.
+
+    :param:        collections    array containing the collections to be
+                                  combined
+    */
+    public init(_ collections: [C]) {
+        self.pools   = collections
+        self.indices = collections.map{ $0.startIndex }
+        self.done    = pools.reduce(true, { $0 && $1.count == 0 })
```

Comment:
While this is nice and compact, it doesn't make the purpose of the line evident at a glance.
```swift
pools.reduce(true, { $0 && $1.count == 0 })
```
Perhaps it could be replaced by something like:
```swift
pools.filter({ !$0.isEmpty }).isEmpty
```
or even a more verbose, but perhaps more readable version:
```swift
let poolsWithContent = pools.filter({ !$0.isEmpty })

self.done = poolsWithContent.isEmpty
```